README.md
# joda-time-fake [![Build Status](https://travis-ci.org/shibayu36/joda-time-fake-scala.svg?branch=master)](https://travis-ci.org/shibayu36/joda-time-fake-scala)

Provides utilities to fake current time gotten by [joda-time](http://www.joda.org/joda-time/).  This is useful for test.

## Getting started

Add dependency in your `build.sbt` as the following.

```scala
libraryDependencies ++= Seq(
  "com.github.shibayu36" %% "joda-time-fake" % "0.0.2"
)
```

The library is available on [Maven Central](https://maven-badges.herokuapp.com/maven-central/com.github.shibayu36/joda-time-fake_2.12).  Currently,
supported Scala versions are 2.11 and 2.12.

## Basic usage

```scala
import com.github.shibayu36.jodatimefake.FakeTimer
import org.joda.time.DateTime

// fake by millis
val result = FakeTimer.fake(1515974400000L) {
    println(DateTime.now.toString) // 2018-01-15T00:00:00.000Z
    "hoge"
}

// current time is restored after block
println(DateTime.now.toString) // current time is printed

// You can get the value returned from block
println(result) // hoge

// You can also pass DateTime instance
FakeTimer.fake(new DateTime(2018, 2, 13, 14, 59)) {

}

// You can also pass ISODateTimeFormat
FakeTimer.fake("2018-03-02T12:34:56+09:00") {

}

// If you use fakeWithTimer, a timer instance is passed to block.
// You can advance time using tick() method.
FakeTimer.fakeWithTimer(1515974400000L) { t =>
    println(DateTime.now.toString) // 2018-01-15T00:00:00.000Z
    t.tick(3000) // Advance time by 3000ms
    println(DateTime.now.toString) // 2018-01-15T00:00:03.000Z
}
```

## Usage in parallel test by thread

By extending `ExecutionContext` with `withFakeTimer` method, you can use `FakeTimer` in parallel test.

```scala
import com.github.shibayu36.jodatimefake.FakeTimer
import com.github.shibayu36.jodatimefake.Implicits._
import org.joda.time.DateTime
import java.util.concurrent.Executors
import scala.concurrent._

// If you use in parallel test, extend ExecutionContext with withFakeTimer method
implicit val ec = ExecutionContext.fromExecutorService(Executors.newFixedThreadPool(3)).withFakeTimer

// The usage is the same as in Basic Usage
val f1 = Future {
    // fake by millis
    val result = FakeTimer.fake(1515974400000L) {
        println(DateTime.now.toString) // 2018-01-15T00:00:00.000Z
        "hoge"
    }

    // current time is restored after block
    println(DateTime.now.toString) // current time is printed

    // You can get the value returned from block
    println(result) // hoge
}

val f2 = Future {
    FakeTimer.fake(1515974400000L) {
        println(DateTime.now.toString) // 2018-01-15T00:00:00.000Z
        t.tick(3000) // Advance time by 3000ms
        val f = Future {
            // Time is inherited from the parent thread
            println(DateTime.now.toString) // 2018-01-15T00:00:03.000Z
        }
        Await.result(f, duration.Duration.Inf)
        println(DateTime.now.toString) // 2018-01-15T00:00:03.000Z
    }
}

println(DateTime.now.toString) // current time is printed
Await.result(f1 zip f2, duration.Duration.Inf)
```


CHANGELOG.md
# Changelog

## [v0.1.0](https://github.com/shibayu36/joda-time-fake-scala/compare/v0.0.3...v0.1.0) (2020-08-04)

* Add support for testing in multi-threaded [#4](https://github.com/shibayu36/joda-time-fake-scala/pull/4) ([pyto86pri](https://github.com/pyto86pri))

## v0.0.3 (2020-07-17)

* Support for Scala 2.13

## v0.0.2 (2018-03-21)

* The feature has not changed
* Add scaladoc for user to understand basic usage
* Add compilation options to check more warnings

## v0.0.1 (2018-03-17)

* Initial release


src/test/scala/org/shibayu36/jodatimefake/FakeTimerSpec.scala
package org.shibayu36.jodatimefake

import org.scalatest.{FunSpec, Matchers}
import org.joda.time.{DateTime, DateTimeZone}

class FakeTimerSpec extends FunSpec with Matchers {
  val tz = DateTimeZone.forID("Asia/Tokyo")

  describe("FakeTimer.fake") {
    it("Fix time only in the block") {
      FakeTimer.fake(1515974400000L) {
        DateTime.now(tz).toString shouldBe "2018-01-15T09:00:00.000+09:00"
      }

      // Restore after block
      DateTime.now(tz).toString shouldNot be("2018-01-15T09:00:00.000+09:00")
    }

    it("Returns value from block") {
      val num = FakeTimer.fake(1515974400000L) {
        123
      }
      num shouldBe 123

      val str = FakeTimer.fake(1515974400000L) {
        "hoge"
      }
      str shouldBe "hoge"
    }

    it("DateTime object can be passed") {
      val dt = new DateTime(2018, 2, 13, 14, 59, tz)
      FakeTimer.fake(dt) {
        DateTime.now(tz).toString shouldBe "2018-02-13T14:59:00.000+09:00"
      }
    }


    it("ISODateTimeFormat string can be passed") {
      FakeTimer.fake("2018-03-02T12:34:56+09:00") {
        DateTime.now(tz).toString shouldBe "2018-03-02T12:34:56.000+09:00"
      }
    }
  }

  describe("FakeTimer.fakeWithTimer") {
    it("Fake time by millis with timer") {
      val result = FakeTimer.fakeWithTimer(1515974400000L) { t =>
        DateTime.now(tz).toString shouldBe "2018-01-15T09:00:00.000+09:00"
        t.tick(100)
        DateTime.now(tz).toString shouldBe "2018-01-15T09:00:00.100+09:00"
        t.tick(1000)
        DateTime.now(tz).toString shouldBe "2018-01-15T09:00:01.100+09:00"
        t.tick(3600 * 1000)
        DateTime.now(tz).toString shouldBe "2018-01-15T10:00:01.100+09:00"

        DateTime.now(tz).toString
      }

      // Last value of block is returned
      result shouldBe "2018-01-15T10:00:01.100+09:00"

      // Restore after block
      DateTime.now(tz).toString shouldNot be(result)
    }

    it("Fake time by DateTime with timer") {
      val dt = new DateTime(2018, 2, 13, 14, 59, tz)
      FakeTimer.fakeWithTimer(dt) { t =>
        DateTime.now(tz).toString shouldBe "2018-02-13T14:59:00.000+09:00"
        t.tick(2000)
        DateTime.now(tz).toString shouldBe "2018-02-13T14:59:02.000+09:00"
      }
    }

    it("Fake time by ISODateTimeFormat string with timer") {
      FakeTimer.fakeWithTimer("2018-03-02T12:34:56+09:00") { t =>
        DateTime.now(tz).toString shouldBe "2018-03-02T12:34:56.000+09:00"
        t.tick(60 * 1000)
        DateTime.now(tz).toString shouldBe "2018-03-02T12:35:56.000+09:00"
      }
    }
  }
}


src/test/scala/com/github/shibayu36/jodatimefake/FakeTimerSpec.scala
package com.github.shibayu36.jodatimefake

import org.scalatest.funspec.AnyFunSpec
import org.scalatest.matchers.should.Matchers

import org.joda.time.{DateTime, DateTimeZone}

import scala.concurrent._

class FakeTimerSpec extends AnyFunSpec with Matchers {
  val tz = DateTimeZone.forID("Asia/Tokyo")

  describe("FakeTimer.fake") {
    it("Fix time only in the block") {
      FakeTimer.fake(1515974400000L) {
        DateTime.now(tz).toString shouldBe "2018-01-15T09:00:00.000+09:00"
      }

      // Restore after block
      DateTime.now(tz).toString shouldNot be("2018-01-15T09:00:00.000+09:00")
    }

    it("Returns value from block") {
      val num = FakeTimer.fake(1515974400000L) {
        123
      }
      num shouldBe 123

      val str = FakeTimer.fake(1515974400000L) {
        "hoge"
      }
      str shouldBe "hoge"
    }

    it("DateTime object can be passed") {
      val dt = new DateTime(2018, 2, 13, 14, 59, tz)
      FakeTimer.fake(dt) {
        DateTime.now(tz).toString shouldBe "2018-02-13T14:59:00.000+09:00"
      }
    }


    it("ISODateTimeFormat string can be passed") {
      FakeTimer.fake("2018-03-02T12:34:56+09:00") {
        DateTime.now(tz).toString shouldBe "2018-03-02T12:34:56.000+09:00"
      }
    }
  }

  describe("FakeTimer.fakeWithTimer") {
    it("Fake time by millis with timer") {
      val result = FakeTimer.fakeWithTimer(1515974400000L) { t =>
        DateTime.now(tz).toString shouldBe "2018-01-15T09:00:00.000+09:00"
        t.tick(100)
        DateTime.now(tz).toString shouldBe "2018-01-15T09:00:00.100+09:00"
        t.tick(1000)
        DateTime.now(tz).toString shouldBe "2018-01-15T09:00:01.100+09:00"
        t.tick(3600 * 1000)
        DateTime.now(tz).toString shouldBe "2018-01-15T10:00:01.100+09:00"

        DateTime.now(tz).toString
      }

      // Last value of block is returned
      result shouldBe "2018-01-15T10:00:01.100+09:00"

      // Restore after block
      DateTime.now(tz).toString shouldNot be(result)
    }

    it("Fake time by DateTime with timer") {
      val dt = new DateTime(2018, 2, 13, 14, 59, tz)
      FakeTimer.fakeWithTimer(dt) { t =>
        DateTime.now(tz).toString shouldBe "2018-02-13T14:59:00.000+09:00"
        t.tick(2000)
        DateTime.now(tz).toString shouldBe "2018-02-13T14:59:02.000+09:00"
      }
    }

    it("Fake time by ISODateTimeFormat string with timer") {
      FakeTimer.fakeWithTimer("2018-03-02T12:34:56+09:00") { t =>
        DateTime.now(tz).toString shouldBe "2018-03-02T12:34:56.000+09:00"
        t.tick(60 * 1000)
        DateTime.now(tz).toString shouldBe "2018-03-02T12:35:56.000+09:00"
      }
    }
  }

  describe("Parallel execution") {
    import java.util.concurrent.Executors
    import scala.concurrent.ExecutionContext

    import Implicits._

    implicit val ec = ExecutionContext.fromExecutorService(Executors.newFixedThreadPool(3)).withFakeTimer

    describe("FakeTimer.fake") {
      it("Can run in multi-threaded") {
        def runThread(msec: Int) = Future {
          FakeTimer.fake(1515974400000L) {
            Thread.sleep(msec)
            DateTime.now(tz).toString shouldBe "2018-01-15T09:00:00.000+09:00"
          }
        }

        Await.result(runThread(100) zip runThread(200), duration.Duration.Inf)
      }

      it("Can run in nested-threaded") {
        def runThread(msec: Int) = Future {
          FakeTimer.fake(1515974500000L) {
            Thread.sleep(msec)
            DateTime.now(tz).toString shouldBe "2018-01-15T09:01:40.000+09:00"
            val f = Future {
              DateTime.now(tz).toString shouldBe "2018-01-15T09:01:40.000+09:00"
            }
            Await.result(f, duration.Duration.Inf)
            DateTime.now(tz).toString shouldBe "2018-01-15T09:01:40.000+09:00"
          }
        }

        Await.result(runThread(100) zip runThread(200), duration.Duration.Inf)
        Await.result(runThread(100) zip runThread(200), duration.Duration.Inf)
      }
    }

    describe("FakeTimer.fakeWithTimer") {
      it("Can run in multi-threaded") {
        def runThread(msec: Int) = Future {
          Thread.sleep(msec)
          FakeTimer.fakeWithTimer(1515974400000L) { t =>
            DateTime.now(tz).toString shouldBe "2018-01-15T09:00:00.000+09:00"
            t.tick(2000)
            DateTime.now(tz).toString shouldBe "2018-01-15T09:00:02.000+09:00"
          }
        }

        Await.result(runThread(100) zip runThread(200), duration.Duration.Inf)
      }

      it("Can run in nested-threaded") {
        def runThread(msec: Int) = Future {
          Thread.sleep(msec)
          FakeTimer.fakeWithTimer(1515974500000L) { t =>
            DateTime.now(tz).toString shouldBe "2018-01-15T09:01:40.000+09:00"
            t.tick(2000)
            DateTime.now(tz).toString shouldBe "2018-01-15T09:01:42.000+09:00"
            val f = Future {
              DateTime.now(tz).toString shouldBe "2018-01-15T09:01:42.000+09:00"
            }
            Await.result(f, duration.Duration.Inf)
            DateTime.now(tz).toString shouldBe "2018-01-15T09:01:42.000+09:00"
          }
        }

        Await.result(runThread(100) zip runThread(200), duration.Duration.Inf)
        Await.result(runThread(100) zip runThread(200), duration.Duration.Inf)
      }
    }
  }
}


src/main/scala/com/github/shibayu36/jodatimefake/FakeTimer.scala
package com.github.shibayu36.jodatimefake

import org.joda.time.{DateTime, DateTimeUtils}
import scala.util.DynamicVariable
import scala.concurrent.ExecutionContext

/** Provides utility methods to fake current time of org.joda.time.DateTime.
 *
 * ==Provided Methods==
 * <ul>
 *   <li>fake</li>
 *   <li>fakeWithTimer</li>
 * </ul>
 * ==Basic Usage==
 * {{{
 * import com.github.shibayu36.jodatimefake.FakeTimer
 * import org.joda.time.DateTime
 *
 * // fake by millis
 * val result = FakeTimer.fake(1515974400000L) {
 *     println(DateTime.now.toString) // 2018-01-15T00:00:00.000Z
 *     "hoge"
 * }
 *
 * // current time is restored after block
 * println(DateTime.now.toString) // current time is printed
 *
 * // You can get the value returned from block
 * println(result) // hoge
 *
 * // You can also pass DateTime instance
 * FakeTimer.fake(new DateTime(2018, 2, 13, 14, 59)) {
 *
 * }
 *
 * // You can also pass ISODateTimeFormat
 * FakeTimer.fake("2018-03-02T12:34:56+09:00") {
 *
 * }
 *
 * // If you use fakeWithTimer, a timer instance is passed to block.
 * // You can advance time using tick() method.
 * FakeTimer.fakeWithTimer(1515974400000L) { t =>
 *     println(DateTime.now.toString) // 2018-01-15T00:00:00.000Z
 *     t.tick(3000) // Advance time by 3000ms
 *     println(DateTime.now.toString) // 2018-01-15T00:00:03.000Z
 * }
 * }}}
 */
object FakeTimer {
  /** Fakes DateTime.now only in passed block by specified timeMillis.
   *
   * Time will be restored after block.
   */
  def fake[T](timeMillis: Long)(block: => T): T = {
    val timer = new FakeTimer(timeMillis)
    DynamicDateTimeUtils.setCurrentMillisProvider(new FakeTimerMillisProvider(timer))
    try {
      block
    } finally {
      DynamicDateTimeUtils.setCurrentMillisSystem()
    }
  }

  /** Fakes DateTime.now by specified DateTime object. */
  def fake[T](t: DateTime)(block: => T): T =
    fake(t.getMillis)(block)

  /** Fakes DateTime.now by specified ISODateTimeFormat string
   *
   * e.g. TimeFaker.fake("2018-03-01T12:34:56") { }
   */
  def fake[T](t: String)(block: => T): T =
    fake(DateTime.parse(t).getMillis)(block)

  /** Fakes DateTime.now by specified timeMillis.
   *
   * This method passes FakeTimer instance to block,
   * so you can advance time by tick method.
   *
   * {{{
   * FakeTimer.fakeWithTimer(1515974400000L) { t =>
   *   t.tick(3000) // Advance time by 3000ms
   * }
   * }}}
   */
  def fakeWithTimer[T](timeMillis: Long)(block: FakeTimer => T): T = {
    val timer = new FakeTimer(timeMillis)
    DynamicDateTimeUtils.setCurrentMillisProvider(new FakeTimerMillisProvider(timer))
    try {
      block(timer)
    } finally {
      DynamicDateTimeUtils.setCurrentMillisSystem()
    }
  }

  /** DateTime object version of fakeWithTimer */
  def fakeWithTimer[T](t: DateTime)(block: FakeTimer => T): T =
    fakeWithTimer(t.getMillis)(block)

  /** ISODateTimeFormat string version of fakeWithTimer */
  def fakeWithTimer[T](t: String)(block: FakeTimer => T): T =
    fakeWithTimer(DateTime.parse(t).getMillis)(block)
}

/** FakeTimer class to advance time in fakeWithTimer */
class FakeTimer(private[this] var currentMillis: Long) {
  /** Advance time by millis */
  def tick(millis: Long): Unit = {
    currentMillis = currentMillis + millis
  }

  private[jodatimefake] def getMillis(): Long = currentMillis
}

private[jodatimefake] class FakeTimerMillisProvider(timer: FakeTimer) extends DateTimeUtils.MillisProvider {
  def getMillis(): Long = timer.getMillis()
}

private[jodatimefake] object SystemMillisProvider extends DateTimeUtils.MillisProvider {
  def getMillis(): Long = System.currentTimeMillis
}

private[jodatimefake] object DynamicDateTimeUtils {
  def setCurrentMillisSystem() = {
      install()
      DynamicMillisProvider.set(SystemMillisProvider)
  }

  def setCurrentMillisProvider(millisProvider: DateTimeUtils.MillisProvider) = {
      install()
      DynamicMillisProvider.set(millisProvider)
  }

  private def install() = DateTimeUtils.setCurrentMillisProvider(DynamicMillisProvider)
}

private[jodatimefake] object DynamicMillisProvider extends DateTimeUtils.MillisProvider {
  private val local = new DynamicVariable[DateTimeUtils.MillisProvider](SystemMillisProvider)

  def get = local.value

  def set(millisProvider: DateTimeUtils.MillisProvider) = local.value_=(millisProvider)

  override def getMillis(): Long = local.value.getMillis
}

object Implicits {
  implicit class RichExecutionContext(ec: ExecutionContext) {
    def withFakeTimer: ExecutionContext = new ExecutionContext {
      override def execute(task: Runnable): Unit = {
        val copyValue = DynamicMillisProvider.get
        ec.execute(new Runnable {
          override def run = {
            DynamicMillisProvider.set(copyValue)
            task.run
          }
        })
      }

      override def reportFailure(cause: Throwable): Unit = ec.reportFailure(cause)
    }
  }
}
